R/decostand.R

```diff
@@ -269,3 +311,323 @@
+    Z <- tmpgrad$Z
+    # line search for the optimum jump length
+    t <- .aux_getoptT(X, W, Y, Z, S, M_E, E, m0, rho)
+    X <- X + t*W;
+    Y <- Y + t*Z;
+    S <- .aux_getoptS(X, Y, M_E, E)
+    # compute the distortion
+    dist[i+1] <- norm(((M_E - X %*% S %*% t(Y))*E),'f') / sqrt(nnZ.E)
+    if (dist[i+1] < TOL){
+      dist <- dist[1:(i+1)]
+      break
+    }
+  }
+  S <- S/rescal_param  
+  # Return Results
+  out <- list()
+  
+  # re-order Optspace may change order during iters
+  index_order <- order(diag(S), decreasing = TRUE)
+  X <- X[, index_order]
+  Y <- Y[, index_order]
+  S <- S[index_order, index_order]
+  out$X <- X
+  out$S <- S
+  out$Y <- Y
+  out$dist <- dist
+  if (verbose){
+    cat('* OptSpace: estimation finished.')
+  }
+
+  # -------------------------------------------
+
+  # This part is not in the Python/Gemelli implementation
+  # but has been added in R to provide more direct access
+  # to the imputed matrix.
+
+  # Reconstruct the matrix
+  M <- X %*% S %*% t(Y)
+
+  # Centering is common operation supporting output visualization
+  # Center cols to 0
+  M <- as.matrix(scale(M, center=TRUE, scale=FALSE))
+  # Center rows to 0
+  M <- as.matrix(t(scale(t(M), center=TRUE, scale=FALSE)))
+
+  # Imputed matrix; add to the output
+  out$M <- M
+
+  # -------------------------------------------
+
+  return(out)
+}
+
+
+# @keywords internal
+.guess_rank <- function(X, nnz)
+{
+  maxiter <- 10000
+  n <- nrow(X)
+  m <- ncol(X)
+  epsilon <- nnz/sqrt(m*n)
+  svdX <- svd(X)
+  S0 <- svdX$d
+  
+  nsval0 <- length(S0)
+  S1 <- S0[seq_len(nsval0-1)]-S0[seq(2, nsval0)]  
+  nsval1 <- length(S1)
+  if (nsval1 > 10){
+    S1_ <- S1/mean(S1[seq((nsval1-10), nsval1)])
+  } else {
+    S1_ <- S1/mean(S1[seq_len(nsval1)])
+  }
+  r1 <- 0
+  lam <- 0.05
+  
+  itcounter <- 0
+  while (r1<=0){
+    itcounter <- itcounter+1
+    cost <- array(0, c(1, length(S1_)))
+    for (idx in seq_len(length(S1_))) {
+      cost[idx] <- lam*max(S1_[seq(idx, length(S1_))]) + idx
+    }
+    v2 <- min(cost)
+    i2 <- which(cost==v2)
+    if (length(i2)==1){
+      r1 <- i2-1
+    } else {
+      r1 <- max(i2)-1
+    }
+    lam <- lam + 0.05
+    if (itcounter > maxiter){
+      break
+    }
+  }
+  
+  if (itcounter<=maxiter){
+    cost2 <- array(0, c(1, (length(S0)-1)))
+    for (idx in seq_len(length(S0)-1)){
+      cost2[idx] <- (S0[idx+1]+sqrt(idx * epsilon) * S0[1]/epsilon)/S0[idx]
+    }
+    v2 <- min(cost2)
+    i2 <- which(cost2==v2)
+    if (length(i2)==1){
+      r2 <- i2
+    } else {
+      r2 <- max(i2)
+    }
+    
+    if (r1>r2){
+      r <- r1
+    } else {
+      r <- r2
+    }
+    return(r)
+  } else {
+    r <- min(nrow(X), ncol(X))
+  }
+}
+
+
+# Aux 2 : compute the distortion ------------------------------------------
+# @keywords internal
+.aux_G <- function(X, m0, r)
+{
+  z <- rowSums(X^2)/(2*m0*r)
+  y <- exp((z-1)^2) - 1
+  idxfind <- (z < 1)
+  y[idxfind] <- 0
+  out <- sum(y)
+  return(out)
```

Comment:
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.